src/final_chain/final_chain.cpp

```diff
@@ -0,0 +1,440 @@
+            continue;
+          }
+          db_query.append(DB::Columns::executed_transactions, trx_h);
+          db_query.append(DB::Columns::transactions, trx_h);
+        }
+      }
+      auto trx_db_results = db_query.execute(false);
+      to_execute.reserve(unique_trxs.size());
+      for (uint i = 0; i < unique_trxs.size(); ++i) {
+        if (auto has_been_executed = !trx_db_results[0 + i * 2].empty(); has_been_executed) {
+          continue;
+        }
+        // Non-executed trxs
+        auto const& trx = to_execute.emplace_back(vector_ref<string::value_type>(trx_db_results[1 + i * 2]).toBytes(),
+                                                  false, h256(db_query.get_key(1 + i * 2)));
+        if (replay_protection_service_->is_nonce_stale(trx.getSender(), trx.getNonce())) {
+          to_execute.pop_back();
+        }
+      }
+    }
+    transaction_count_hint_ = max(transaction_count_hint_, to_execute.size());
+    auto const& [exec_results, state_root] =
+        state_api.transition_state({new_blk.author, GAS_LIMIT, new_blk.timestamp, BlockHeader::difficulty()},
+                                   to_state_api_transactions(to_execute));
+    TransactionReceipts receipts;
+    receipts.reserve(exec_results.size());
+    gas_t cumulative_gas_used = 0;
+    for (auto const& r : exec_results) {
+      LogEntries logs;
+      logs.reserve(r.Logs.size());
+      for (auto const& l : r.Logs) {
+        logs.push_back({l.Address, l.Topics, l.Data});
+      }
+      receipts.push_back(TransactionReceipt{
+          r.CodeErr.empty() && r.ConsensusErr.empty(),
+          r.GasUsed,
+          cumulative_gas_used += r.GasUsed,
+          move(logs),
+          r.NewContractAddr ? optional(r.NewContractAddr) : nullopt,
+      });
+    }
+    auto blk_header =
+        append_block(batch, new_blk.author, new_blk.timestamp, GAS_LIMIT, state_root, to_execute, receipts);
+    // Update replay protection service, like nonce watermark. Nonce watermark has been disabled
+    replay_protection_service_->update(
+        batch, blk_header->number, util::make_range_view(to_execute).map([](auto const& trx) {
+          return ReplayProtectionService::TransactionInfo{trx.getSender(), trx.getNonce()};
+        }));
+    // Update number of executed DAG blocks and transactions
+    auto num_executed_dag_blk = num_executed_dag_blk_ + new_blk.dag_blk_hashes.size();
+    auto num_executed_trx = num_executed_trx_ + to_execute.size();
+    if (!new_blk.dag_blk_hashes.empty()) {
+      db_->insert(batch, DB::Columns::status, StatusDbField::ExecutedBlkCount, num_executed_dag_blk);
+      db_->insert(batch, DB::Columns::status, StatusDbField::ExecutedTrxCount, num_executed_trx);
+      LOG(log_nf_) << "Executed dag blocks #" << num_executed_dag_blk_ - new_blk.dag_blk_hashes.size() << "-"
+                   << num_executed_dag_blk_ - 1 << " , Transactions count: " << to_execute.size();
+    }
+    auto result = s_ptr(new FinalizationResult{
```

Comment:
do not use "new", use make_shared/unique instead

Reply:
Here I used `new` just because I wanted to use aggregate initialization. make_shared doesn't allow to do it unfortunately. That means I need to define a normal constructor in FinalizationResult, which I wanted to avoid. You may have noticed that I prefer aggregate structs for simple data types. 

I will find a way to avoid `new` there

Reply:
done